lib/ansible/modules/cloud/azure/azure_rm_deployment.py

```diff
@@ -335,30 +311,82 @@
 
 RETURN = '''
 deployment:
-  description: Deployment details
-  type: dict
-  returned: always
-  sample:
-      group_name:
-        description: Name of the resource group
-        type: str
-        returned: always
-      id:
-        description: The Azure ID of the deployment
-        type: str
-        returned: always
-      instances:
-        description: Provides the public IP addresses for each VM instance.
-        type: list
-        returned: always
-      name:
-        description: Name of the deployment
-        type: str
-        returned: always
-      outputs:
-        description: Dictionary of outputs received from the deployment
-        type: dict
-        returned: always
+    description: Deployment details
+    type: complex
+    returned: always
+    contains:
+        group_name:
+            description: Name of the resource group
+            type: str
+            returned: always
+            sample: myResourceGroup
+        id:
+            description: The Azure ID of the deployment
+            type: str
+            returned: always
+            sample: "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/myResourceGroup/providers/Microsoft.Resources/deployments/myD
+                     eployment"
+        instances:
+            description: Provides the public IP addresses for each VM instance.
+            type: list
+            returned: always
```

Comment:
return always? what if my arm template doesn't contain VM?

Reply:
well, field is always returned, it may just be empty []

Reply:
it's confusing, such as public  ip marked as returned always also.

Reply:
yes, the lists are always returned, just may be empty,
maybe we could have:
**returned: always but list may be empty**

or just
**returned when applicable**

Reply:
I actually think "instances" return value should be made obsolete.
it works only with vms.
if somebody uses ARM templates to deploy anything else than VM it's pretty much useless.
we could think of more universal approach.

Usually **outputs** should be sufficient to use created resources. For instance for Windows VM, outputs look like:
```json
            "outputs": {
                "hostname": {
                    "type": "String",
                    "value": "zimsdplxxxx.eastus2.cloudapp.azure.com"
                }
            }
```

So it has sufficient information to use the VM.
Outputs are designed by ARM template authors so it's up to them what's important for the user.